### PR TITLE
Add Assertions to DiscoveryNodes Constructor

### PR DESCRIPTION
**Pull request by original-brownbear**
Adding assertions as requested in #84276.

### CODE REVIEW

**Review by DaveCTurner — APPROVED** [review]
LGTM thanks

**Comment by DaveCTurner** [comment]
Oh wow this was more broken than I thought...

**Comment by original-brownbear** [comment]
> Oh wow this was more broken than I thought...

and not even done yet ... there's so many randomised tests randomly breaking here. I spent all day fixing these in the background. Almost makes me want to postpone this ...

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/rest-compatibility (unrelated)

**Comment by original-brownbear** [comment]
@DaveCTurner alright I think I found all spots. Figured I re-request review since it's 150loc chagnes now (albeit exclusively in tests).

File: server/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodesTests.java [hunk 1]
```diff
@@ -420,8 +420,6 @@ public void testMaxMinNodeVersion() {
                 Version.fromString("1.1.0")
             )
         );
-        discoBuilder.localNodeId("name_1");
-        discoBuilder.masterNodeId("name_2");
         DiscoveryNodes build = discoBuilder.build();
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/node/DiscoveryNodesTests.java:423]: `name` should have been `node` here (not that it really matters)

```suggestion
        discoBuilder.localNodeId("node_" + between(1, 3));
        if (randomBoolean()) {
            discoBuilder.masterNodeId("node_" + between(1, 3));
        }
        DiscoveryNodes build = discoBuilder.build();
```

**Review by DaveCTurner — APPROVED** [review]
LGTM still

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**